lib/SIL/Verifier/ReborrowVerifier.cpp

```diff
@@ -0,0 +1,141 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-reborrow-checker"
+
+#include "ReborrowVerifierPrivate.h"
+#include "swift/SIL/OwnershipUtils.h"
+
+using namespace swift;
+
+bool ReborrowVerifier::checkGuaranteedPhiArgWithoutLifetimeEndingUses(
+    SILPhiArgument *phiArg) {
+  assert(phiArg->getOwnershipKind() == ValueOwnershipKind::Guaranteed);
+  if (!getRequiresLifetimeEndUse(phiArg))
+    return false;
+
+  errorBuilder.handleMalformedSIL([&] {
+    llvm::errs() << "Non trivial values, non address values, and non "
+                    "guaranteed function args must have at least one "
+                    "lifetime ending use?!\n";
+    llvm::errs() << "Value: " << *phiArg << '\n';
+  });
+  return true;
+}
+
+bool ReborrowVerifier::verify() {
+  SmallVector<Operand *, 4> phiArgUses;
+  SmallPtrSet<SILBasicBlock *, 4> visitedBlocks;
+  bool result = false;
+
+  // For all guaranteed phi args and base value pairs in the worklist, verify
+  // that the guaranteed phi arg lies within the lifetime of the base value.
+  while (!reborrowWorklist.empty()) {
+    auto item = reborrowWorklist.pop_back_val();
+    auto *phiArg = item.first;
+    auto baseVal = item.second;
+
+    phiArgUses.clear();
+    visitedBlocks.clear();
+
+    phiArgUses.append(phiArg->getUses().begin(), phiArg->getUses().end());
+
+    LinearLifetimeChecker checker(visitedBlocks, deadEndBlocks);
+    // newErrorBuilder is consumed at the end of the checkValue function.
+    // Copy initial state from errorBuilder everytime
+    LinearLifetimeChecker::ErrorBuilder newErrorBuilder = errorBuilder;
+    // Check if it is okay to have a guaranteed phi arg without lifetime ending
+    // use
+    if (phiArg->getConsumingUses().empty()) {
+      result |= checkGuaranteedPhiArgWithoutLifetimeEndingUses(phiArg);
+    }
+    SmallVector<Operand*, 4> baseValConsumingUses(baseVal->getConsumingUses());
+    // If the baseValue has no consuming uses, there is nothing more to verify
+    if (baseValConsumingUses.empty())
+      continue;
+    auto linearLifetimeResult = checker.checkValue(
+        baseVal, baseValConsumingUses, phiArgUses, newErrorBuilder);
+    result |= linearLifetimeResult.getFoundError();
+  }
+  return result;
+}
+
+void ReborrowVerifier::populateReborrowVerifier(SILValue value,
+    SmallVectorImpl<Operand *> &initialBorrowLifetimeEndingUsers,
+    bool requiresLifetimeEndingUse) {
+  LLVM_DEBUG(llvm::dbgs() << "Populating ReborrowVerifier\n");
+
+  SmallVector<std::tuple<Operand *, SILValue>, 4> worklist;
```

Comment:
Seems like this would be simpler to just do one at a time with a cleanup worklist rather than trying to do it all at one time. Why not do that?

Reply:
I did that now